File: uzum/python/uzum_payment/uzum_auth.py
"""Uzum Bank'ning Basic Auth so'rovini tekshirish.

Payme'ga o'xshab, Uzum Bank ham har bir webhook so'rovini oddiy HTTP Basic
Authentication bilan tasdiqlaydi:

    Authorization: Basic base64("login:parol")

Farqi: Payme'da login doim "Paycom", Uzum'da esa LOGIN HAM, PAROL HAM siz
kabinetda o'zingiz belgilaysiz (ikkalasi ham `.env` da).

Bu faylga tegishingiz shart emas.
"""

from __future__ import annotations

import base64
import binascii
import hmac

from .uzum_config import UzumConfig


def check_auth(authorization_header: str | None, config: UzumConfig) -> bool:
    """`Authorization` sarlavhasini tekshiradi.

    Login va parol vaqt bo'yicha barqaror (timing-safe) solishtiriladi.
    """
    if not authorization_header or not authorization_header.startswith("Basic "):
        return False

    token = authorization_header[len("Basic "):].strip()

    try:
        decoded = base64.b64decode(token, validate=True).decode("utf-8")
    except (binascii.Error, UnicodeDecodeError, ValueError):
        return False

    login, _, password = decoded.partition(":")
    if not password:
        return False

    login_ok = hmac.compare_digest(login.encode(), config.webhook_login.encode())
    password_ok = hmac.compare_digest(password.encode(), config.webhook_secret.encode())

    return login_ok and password_ok

File: uzum/typescript/uzum-payment/uzum-auth.ts
/**
 * Uzum Bank'ning Basic Auth so'rovini tekshirish.
 *
 * Payme'ga o'xshab, Uzum Bank ham har bir webhook so'rovini oddiy HTTP Basic
 * Authentication bilan tasdiqlaydi:
 *
 *     Authorization: Basic base64("login:parol")
 *
 * Farqi: Payme'da login doim "Paycom", Uzum'da esa LOGIN HAM, PAROL HAM siz
 * kabinetda o'zingiz belgilaysiz.
 *
 * ⚠️ Bu modul `node:crypto` ga tayanadi — Next.js Edge runtime'da
 * ishlamaydi (`export const runtime = "nodejs"` yozing).
 *
 * Bu faylga tegishingiz shart emas.
 */

import { timingSafeEqual } from "node:crypto";

import type { UzumConfig } from "./uzum-config.js";

function safeEqual(a: string, b: string): boolean {
  const bufA = Buffer.from(a, "utf8");
  const bufB = Buffer.from(b, "utf8");
  if (bufA.length !== bufB.length) return false;
  return timingSafeEqual(bufA, bufB);
}

/**
 * `Authorization` sarlavhasini tekshiradi.
 *
 * Login va parol vaqt bo'yicha barqaror (timing-safe) solishtiriladi.
 */
export function checkAuth(
  authorizationHeader: string | null | undefined,
  config: UzumConfig,
): boolean {
  if (!authorizationHeader || !authorizationHeader.startsWith("Basic ")) {
    return false;
  }

  const token = authorizationHeader.slice("Basic ".length).trim();

  let decoded: string;
  try {
    decoded = Buffer.from(token, "base64").toString("utf8");
  } catch {
    return false;
  }

  const sepIndex = decoded.indexOf(":");
  if (sepIndex === -1) return false;

  const login = decoded.slice(0, sepIndex);
  const password = decoded.slice(sepIndex + 1);
  if (!password) return false;

  return safeEqual(login, config.webhookLogin) && safeEqual(password, config.webhookSecret);
}
